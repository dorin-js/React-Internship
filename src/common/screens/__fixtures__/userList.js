export const usersList = [
  {
    _created: 1668698852.871516,
    _data_type: 'users',
    _is_deleted: false,
    _modified: 1668698852.871524,
    _self_link: 'https://crudapi.co.uk/api/v1/users/fde94e38-a0e8-4244-bbce-65943235bb47',
    _user: 'ca1db56c-83bf-4fca-b2e8-6c31fd33f6ff',
    _uuid: 'fde94e38-a0e8-4244-bbce-65943235bb47',
    birth: '2022-11-28',
    email: 'dorin.ilusca@hotmail.com',
    lastname: 'ggg',
    name: '6346366',
  },
  {
    _created: 1668511679.512981,
    _data_type: 'users',
    _is_deleted: false,
    _modified: 1668511679.512994,
    _self_link: 'https://crudapi.co.uk/api/v1/users/9d549a7f-f881-47fb-b615-16fd0b81078e',
    _user: 'ca1db56c-83bf-4fca-b2e8-6c31fd33f6ff',
    _uuid: '9d549a7f-f881-47fb-b615-16fd0b81078e',
    birth: '2022-09-27',
    email: 'ilusca97@gmail.com',
    lastname: 'ILUSCA',
    name: 'DORIN',
  },
];
