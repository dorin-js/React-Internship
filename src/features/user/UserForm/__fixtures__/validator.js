/* eslint-disable no-useless-escape */
export default {
  firstname: {
    required: {
      value: true,
      message: 'Input is required',
    },
    maxLength: {
      value: 30,
      message: 'The input is too long',
    },
    pattern: {
      value: /^[a-zA-Z]+$/,
      message: 'Name should contain letters only',
    },
  },
  lastname: {
    required: {
      value: true,
      message: 'Input is required',
    },
    maxLength: {
      value: 30,
      message: 'The input is too long',
    },
    pattern: {
      value: /^[a-zA-Z]+$/,
      message: 'Name should contain letters only',
    },
  },
  email: {
    required: {
      value: true,
      message: 'Email is required',
    },
    maxLength: {
      value: 30,
      message: 'The input is too long',
    },
    pattern: {
      value: /^\w+([\.-]?\w+)*@\w+([\.]\w+)(\.\w{2,5})?$/,
      message: 'Incorrect email format',
    },
  },
  birthDate: {
    required: {
      value: true,
      message: 'Birth date is required',
    },
  },
};
