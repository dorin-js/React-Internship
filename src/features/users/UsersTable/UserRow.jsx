import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useDeleteUserMutation } from '../../../services/api/apiService';
import { Button } from '../../../common/components/Button';
import { Snackbar } from '../../../common/components/Snackbar';
import Portal from '../../../common/components/Portal';
import Modal from '../../../common/components/Modal/Modal';
import UserForm from '../../user/UserForm/UserForm';

const UserRow = ({ user, onUserDetails }) => {
  const [isEditing, setIsEditing] = useState(false);

  const {
    id, name, lastname, email, birth,
  } = user;

  const [deleteUser, {
    isLoading: isDeleteInProgress,
    isError: isErrorOnDeletingUser,
    error: deleteUserError,
  }] = useDeleteUserMutation();

  const onUpdateUser = () => {
    setIsEditing(true);
  };

  return (
    <>
      <tr data-testid={`row-${name}`}>
        <td>{name}</td>
        <td>{lastname}</td>
        <td>{email}</td>
        <td>{birth}</td>
        <td>
          <div className="buttonsContainer">
            <Button
              data-testid={`${id}-details`}
              value="Show Details"
              onClick={() => onUserDetails(user)}
            />
            <Button
              data-testid={`${id}-delete`}
              value={isDeleteInProgress ? 'Deleting...' : 'Delete'}
              onClick={() => deleteUser(id)}
              disabled={isDeleteInProgress}
            />
            <Button
              value="Edit"
              onClick={onUpdateUser}
              disabled={isEditing}
            />
          </div>
        </td>
      </tr>
      {
        isErrorOnDeletingUser && (
          <Portal>
            <Snackbar type="error" message={deleteUserError.data.error} timeout={4000} />
          </Portal>
        )
      }
      {
        isEditing && (
          <Portal>
            <Modal title="Update user" onClose={() => setIsEditing(false)}>
              <UserForm user={user} isEditing={isEditing} />
            </Modal>
          </Portal>
        )
      }

    </>
  );
};

UserRow.propTypes = {
  user: PropTypes.PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    lastname: PropTypes.string,
    email: PropTypes.string,
    birth: PropTypes.string,
  }),
  onUserDetails: PropTypes.func,
};
UserRow.defaultProps = {
  user: {},
  onUserDetails: undefined,
};

export default UserRow;
