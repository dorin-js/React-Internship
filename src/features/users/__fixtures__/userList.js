export const usersList2 = [
  {
    id: 'e512ee0a-0022-4041-ad26-479e425cbaab',
    email: 'jake@mail.com',
    birth: '2022-10-24',
    lastname: 'Sully',
    name: 'Jake',
  },
  {
    id: 'd31330d2-8d7c-4d63-abcd-bbfa0b7c705f',
    email: 'jonhndoe@mail.co',
    birth: '2022-11-16',
    lastname: 'Doe',
    name: 'John',
  },
];
export const usersList = [
  {
    _created: 1669148545.821376,
    _data_type: 'users',
    _is_deleted: false,
    _modified: 1669148545.821387,
    _self_link: 'https://crudapi.co.uk/api/v1/users/e512ee0a-0022-4041-ad26-479e425cbaab',
    _user: 'ca1db56c-83bf-4fca-b2e8-6c31fd33f6ff',
    _uuid: 'e512ee0a-0022-4041-ad26-479e425cbaab',
    id: 'e512ee0a-0022-4041-ad26-479e425cbaab',
    birth: '2022-10-24',
    email: 'jake@mail.com',
    lastname: 'Sully',
    name: 'Jake',
  },
  {
    _created: 1669146430.114581,
    _data_type: 'users',
    _is_deleted: false,
    _modified: 1669371087.72434,
    _self_link: 'https://crudapi.co.uk/api/v1/users/d31330d2-8d7c-4d63-abcd-bbfa0b7c705f',
    _user: 'ca1db56c-83bf-4fca-b2e8-6c31fd33f6ff',
    _uuid: 'd31330d2-8d7c-4d63-abcd-bbfa0b7c705f',
    birth: '2022-11-16',
    email: 'jonhndoe@mail.co',
    id: 'd31330d2-8d7c-4d63-abcd-bbfa0b7c705f',
    lastname: 'Doe',
    name: 'John',
  },
];
