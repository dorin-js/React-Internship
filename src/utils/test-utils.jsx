import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { PropTypes } from 'prop-types';
import { setupStore } from '../app/store';

export const renderWithProviders = (
  ui,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
  } = {},
) => {
  const Wrapper = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  );

  Wrapper.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return { store, ...render(ui, { wrapper: Wrapper }) };
};
