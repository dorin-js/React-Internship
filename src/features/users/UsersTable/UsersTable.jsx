/* eslint-disable no-underscore-dangle */
import React from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import classes from './UsersTable.module.css';
import { Modal } from '../../../common/components/Modal';
import UserRow from './UserRow';
import { getUserDetails, setDetails } from '../usersSlice';

const UsersTable = ({ users }) => {
  const details = useSelector(getUserDetails);
  const dispatch = useDispatch();

  const onUserDetails = (userData) => {
    dispatch(setDetails(userData));
  };
  return (
    <main className={classes.main}>
      <table className={`${classes.usersTable} ${classes.hscroll}`}>
        <thead>
          <tr>
            <th>First Name</th>
            <th>Last Name</th>
            <th>Email</th>
            <th>Birth Date</th>
          </tr>
        </thead>
        <tbody>
          {users?.map((user) => (
            <UserRow
              key={user.id}
              user={user}
              onUserDetails={onUserDetails}
            />
          ))}
        </tbody>
      </table>

      {details && (
        <Modal onClose={() => onUserDetails(null)}>
          <h5>{details.firstname}</h5>
          <h5>{details.lastname}</h5>
          <h5>{details.email}</h5>
          <h5>{details.birthDate}</h5>
        </Modal>
      )}
    </main>
  );
};

UsersTable.propTypes = {
  users: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    lastname: PropTypes.string,
    email: PropTypes.string,
    birth: PropTypes.string,
  })),
};
UsersTable.defaultProps = {
  users: undefined,
};

export default UsersTable;
