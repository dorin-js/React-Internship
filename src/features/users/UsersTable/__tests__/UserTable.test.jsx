import React from 'react';
import { usersList } from '../../__fixtures__/userList';
import { renderWithProviders } from '../../../../utils/test-utils';
import UsersTable from '../UsersTable';

describe('Users Table', () => {

  it('should render table with 2 users', async () => {
    const { baseElement, getByTestId, getAllByTestId } = renderWithProviders(
      <UsersTable users={usersList} />
    );

    expect(baseElement).toMatchSnapshot();

    expect(getByTestId('row-John')).toBeInTheDocument();

    expect(getAllByTestId(/row/)).toHaveLength(usersList.length);

    // fireEvent.click(getByRole(('button', { name: 'Show Details'})));
    // await waitFor(() => {
    //   expect().toHaveBeenCalledWith(usersList[1]);
    // });
  });
})