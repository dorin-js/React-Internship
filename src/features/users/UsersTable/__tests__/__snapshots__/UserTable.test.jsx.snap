// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Table should render table with 2 users 1`] = `
<body>
  <div>
    <div
      id="portal-root"
    >
      <main
        class="main"
      >
        <table
          class="usersTable"
        >
          <thead>
            <tr>
              <th>
                First Name
              </th>
              <th>
                Last Name
              </th>
              <th>
                Email
              </th>
              <th>
                Birth Date
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              data-testid="row-Jake"
            >
              <td>
                Jake
              </td>
              <td>
                Sully
              </td>
              <td>
                jake@mail.com
              </td>
              <td>
                2022-10-24
              </td>
              <td>
                <div
                  class="buttonsContainer"
                >
                  <button
                    class="button"
                    data-testid="e512ee0a-0022-4041-ad26-479e425cbaab-details"
                    name="Show Details"
                    type="button"
                  >
                    <span
                      class="buttonText"
                    >
                      Show Details
                    </span>
                  </button>
                  <button
                    class="button"
                    data-testid="e512ee0a-0022-4041-ad26-479e425cbaab-delete"
                    name="Delete"
                    type="button"
                  >
                    <span
                      class="buttonText"
                    >
                      Delete
                    </span>
                  </button>
                  <button
                    class="button"
                    name="Edit"
                    type="button"
                  >
                    <span
                      class="buttonText"
                    >
                      Edit
                    </span>
                  </button>
                </div>
              </td>
            </tr>
            <tr
              data-testid="row-John"
            >
              <td>
                John
              </td>
              <td>
                Doe
              </td>
              <td>
                jonhndoe@mail.co
              </td>
              <td>
                2022-11-16
              </td>
              <td>
                <div
                  class="buttonsContainer"
                >
                  <button
                    class="button"
                    data-testid="d31330d2-8d7c-4d63-abcd-bbfa0b7c705f-details"
                    name="Show Details"
                    type="button"
                  >
                    <span
                      class="buttonText"
                    >
                      Show Details
                    </span>
                  </button>
                  <button
                    class="button"
                    data-testid="d31330d2-8d7c-4d63-abcd-bbfa0b7c705f-delete"
                    name="Delete"
                    type="button"
                  >
                    <span
                      class="buttonText"
                    >
                      Delete
                    </span>
                  </button>
                  <button
                    class="button"
                    name="Edit"
                    type="button"
                  >
                    <span
                      class="buttonText"
                    >
                      Edit
                    </span>
                  </button>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </main>
    </div>
  </div>
</body>
`;
